p2p/src/interface/msg_channel.rs

```diff
@@ -0,0 +1,180 @@
+        match config {
+            // Do not use a buffer, instead block according to the Sink implementation for the inner `mpsc::Sender`.
+            ChannelSinkConfig::Block => {
+                let (tx, rx) = mpsc::channel(capacity);
+                (
+                    EventChannel {
+                        inner: tx,
+                        buffer: None,
+                        use_inner: true,
+                        waker: None,
+                    },
+                    rx,
+                )
+            }
+            // Do not use a buffer, drop new events if `mpsc::Sender::try_send` failed due to a full channel.
+            ChannelSinkConfig::DropLatest => {
+                let (inner, rx) = mpsc::channel(capacity);
+                (
+                    EventChannel {
+                        inner,
+                        buffer: None,
+                        use_inner: false,
+                        waker: None,
+                    },
+                    rx,
+                )
+            }
+            // Use a buffer for the latest events if `mpsc::Sender::try_send` failed due to a full channel.
+            ChannelSinkConfig::BufferLatest => {
+                // Use capacity of 1 since the mpsc::channel only stores the first n events, rather then the last.
+                // Instead use a ring-buffer with the set capacity to buffer the most recent n events.
+                let (inner, rx) = mpsc::channel(1);
+                (
+                    EventChannel {
+                        inner,
+                        buffer: Some(VecDeque::with_capacity(capacity)),
+                        use_inner: false,
+                        waker: None,
+                    },
+                    rx,
+                )
+            }
+        }
+    }
+}
+
+impl<T> Unpin for EventChannel<T> {}
+
+/// Implement [`Sink`] for sending events through the underlying channel.
+impl<T> Sink<T> for EventChannel<T> {
+    type Error = <mpsc::Sender<T> as Sink<T>>::Error;
+
+    fn poll_ready(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        if self.use_inner {
+            self.inner.poll_ready(cx)
+        } else {
+            Poll::Ready(Ok(()))
+        }
+    }
+
+    fn start_send(mut self: Pin<&mut Self>, item: T) -> Result<(), Self::Error> {
+        if self.use_inner {
+            let res = self.inner.start_send(item);
+            return res;
+        }
+        match self.inner.try_send(item) {
+            Ok(()) => Ok(()),
+            Err(e) if e.is_full() => match self.buffer.as_mut() {
+                Some(buffer) => {
+                    if buffer.len() >= buffer.capacity() {
+                        // Drop older events in favor of new ones.
+                        buffer.pop_front();
+                    }
+                    buffer.push_back(e.into_inner());
+                    if let Some(waker) = self.waker.as_ref() {
```

Comment:
```suggestion
                    if let Some(waker) = self.waker.take() {
```

I am not 100% sure, though I would expect to be woken up at most once when waiting for a new item to be added to the `buffer`, no?

Reply:
I actually had it like this at first, but because I was also not 100% sure I changed it to not consume the waker. 

But you are right. If a new item was added to the buffer and the task was waken again, it will attempt to send the request. Thus wakening the task is then related to the channel being ready, and not to new items in the buffer.